doc/man/tlstunnel.man
.TH TLSTUNNEL 8 "November 2015" "ARPA2.net" "System Management Commands"
.SH NAME
tlstunnel \- Forward TLS/DTLS as TCP/UDP/SCTP, and vice versa.
.SH SYNOPSIS
.B tlstunnel
.IR MODE " " OPTIONS " [" -- "] " <pavlov_cmd>...
.SH DESCRIPTION
.PP
The
.B tlstunnel
command wraps TLS around connections that are otherwise plaintext.
This can even be used for protocols that do STARTTLS in the
course of their protocol interactions, due to a simple
.IR pavlov (3)
scripting facility.
.PP
The TLS facilities are drawn from a
.IR tlspool-daemon (8)
daemon, which is contacted over a UNIX domain socket.  It therefore
inherits the architectural benefits of the TLS Pool, including the
ability of central control, strong protection of cryptographic
credentials, and even the ability to monitor and control which local
identities are revealed to which remote peers.
.PP
The tunnel can be built up using either TLS or DTLS, the latter of which
is the datagram variety of common, stream-oriented TLS.  When relaying
a TCP connection, TLS is the only available option; when relaying a
UDP connection, DTLS is the only available option; when relaying an
SCTP connection, both varieties are available but DTLS is generally
preferred for efficiency reasons.
.PP
Future versions of
.B tlstunnel
are expected to support symmetric behaviour, but the present version
must run in either client or server mode.  Used as a client, the
service listens for plaintext connections made to address \fIinaddr\fR and
port \fIinport\fR, wraps them in TLS as a client and forwards the connection
to address \fIfwaddr\fR and port \fIfwport\fR.  When used as a server, the same
is done, however the wrapping of TLS is changed to unwrapping of TLS.
.PP
Addresses can take the form of a UNIX domain socket path, or host:port where
the host is either a DNS name or an IPv6 address.  For the last form,
backward compatibility to IPv4 is possible by prefixing :: to an IPv4 address.
It is possible to use this backward compatibility with either or both or none
of \fIinaddr\fR and \fIfwaddr\fR.
.PP
The end points authenticate, and are authorized, as configured in the
system-wide TLS Pool policy settings.  These settings ensure a
consistent default behaviour for all TLS Pool connections.  As an example,
these policies tend to include requirements to certificates.
.PP
The \fIMODE\fR defines whether \fItlstunnel\fR runs as a client or server:
.TP
.BR -c ", " --client
Run
.B tlstunnel
as a client, meaning that it is connected to by
a plaintext connection and forwarded as a TLS-wrapped variety.

Client authentication and authorization are implemented as specified
in the TLS Pool configuration.  This includes choice of certificates
and settings about their visibility.
.TP
.BR -s ", " --server
Run
.B tlstunnel
as a server, meaning that it is connected to by
a TLS-wrapped connection and forwarded as a plaintext variety.

Client authentication and authorization are implemented as specified
in the TLS Pool configuration.  This includes choice of certificates
and settings about their visibility.
.PP
The \fIOPTIONS\fR detail further information, some of which is
actually not optional at all.
.TP
.BR -f ", " --fork
When the
.B tlstunnel
stops, the TLS connections should continue to run.
.TP
.BR -t ", " --tcp ", " --tcp-tls
Listen to, and forward as, TCP connection.  The secure wrapping
is implemented through stream-oriented TLS.

This is also the default option, selected when none of
.BR --tcp-tls ", " --udp-dtls ", " --sctp-tls " or " --sctp-dtls
has been supplied.
.TP
.BR -u ", " --udp ", " --udp-dtls
THIS FEATURE IS NOT IMPLEMENTED.  The semantics cannot be clearly
defined, because UDP has no notion of a connection, while TLS does
need it to keep state.  Rather than implementing guesses here and
load it with configuration options such as timeouts, it is advised
to implement DTLS directly in the UDP application.
.TP
.BR "-x \fIstream\fB" ", " "--sctp-stream-number \fIstream\fB"
When SCTP is used, forward the given stream number.  This option may
in the future be REDESIGNED to permit multiple streams, and the default
is unsettled but might be all.  There are a few alternative approaches.

Combine with one of
.BR --sctp-dtls " or " --sctp-tls
to specify the desired form of transport layer security.
.TP
.B --sctp-tunneled
Specifies that SCTP should be tunneled over UDP, using the
port number 9899 as specified in RFC 6951.  The default is to not use
a tunnel, as can be specified explicitly with
.BR --sctp-not-tunneled .
.TP
.B --sctp-not-tunneled
Specifies that SCTP should not be tunneled over UDP.  This is the
default behaviour when \fI--sctp-tunneled\fR is not specified.
.TP
.BR "-l \fIlocaladdr\fB" ", " "--local-addr \fIlocaladdr\fB"
The local address for the tunnel.
.TP
.BR "-r \fIremoteaddr\fB" ", " "--remote-addr \fIremoteaddr\fB"
The remote address for the tunnel.
.TP
.BR "-L \fIlid\fB" ", " "--local-id \fIlid\fB"
Set the local identity.  This should be of the form of a domain
name, or user@domainname.  This paramater is required.  When a
localid service program has registered with
.IR tlspool-daemon (8)
then the local identity may be passed through for filtering, depending
on the option used when registering.
.TP
.BR "-R \fIrid\fB" ", " "--remote-id \fIrid\fB"
Constrain the remote identity.  This should be of the form of a domain
name, or user@domainname.  This parameter is optional in server mode,
but is (usually) needed to setup client mode.
.TP
.BR "-s \fIsvcname\fB" ", " "--service \fIsvcname\fB"
Specify the service name that will be sent to
.IR tlspool-daemon (8)
during initiation of the TLS protection.  This helps to decide several
protocol-specifics, including whether it is safe to start with an
Anonymising Precursor.  The service name should follow the IANA spelling,
which is also used in
.IR /etc/services .
.TP
.BR "-S \fIpath\fB" ", " "--tlspool-socket-path \fIpath\fB"
Override the built-in default path for the socket pool.  This can be
used to address a specific instance of the TLS Pool, for instance one
that is run under a personal account, accessing a personally held
PKCS #11 token.
.BR "-p command" ", "--pavlov-command command"
Specifies an alternative command in the style of
.IR pavlov (3)
to replace the builtin version.
An external command will normally be less efficient than the
builtin variety.
.PP
The remaining commandline arguments form the \fI<pavlov_cmd>...\fR, which are passed
on to the builtin or external
.IR pavlov (3)
facility.  This can be used to perform communication in plaintext over
the soon-to-be TLS connection; it can be used to negotiate options over
a textual protocol up to the agreement to switch to TLS.  This is
precisely what
.IR pavlov (3)
was design for; except that the next-stage protocol was PPP and not TLS.
.SH LIMITATIONS
.PP
The TLS Pool can handle various options when connecting to a remote site.
These options have not yet been implemented as commandline switches, so
this level of refinement cannot be added when using the 
.B tlstunnel
program.
.PP
The TLS Pool usually negotiates local and remote identities as part of
the authentication and authorization procedures.  These procedures do not
interface optimally with the TLS Tunnel, because there is no place to
leave this information.  What has been implemented is the mandatory
-l option that sets the local identifier to use, and an optional -r option
to constrain the remote identity.
.PP
Note that the TLS Tunnel does not implement a PIN entry interface; this
is a function that is separately coordinated with the TLS Pool, so it can
bring up a dialog in a trusted environment, making it easier to secure than
any per-command PIN entry dialog.
.PP
Where TLS and even DTLS have a notion of connection state, necessary for the
management of things like session keys, there is no concept of a connection
for UDP transmission.  For this reason, the UDP variant of this tool has not
been implemented; it would involve close-to-NAT practices that can easily
lead to confusion.
.SH AUTHOR
.PP
Written by Rick van Rein of OpenFortress.nl, for the ARPA2.net project.
.SH "REPORTING BUGS"
.PP
For any discussion, including about bugs, please use the TLS Pool
mailing list found on
.IR http://lists.arpa2.org/mailman/listinfo/tls-pool .
.PP
Please read the software distribution's
.IR README ", " INSTALL " and " TODO " files"
for information about the
.IR tlspool-daemon (8)
implementation status.
.SH COPYRIGHT
.PP
Copyright \(co 2015 Rick van Rein, ARPA2.net.
.PP
ARPA2 is funded from InternetWide.org, which in turns receives donations
from various funding sources with an interest in a private and secure
Internet that gives users control over their online presence.  This particular
project has been sponsored in part by NCSC.
.SH "SEE ALSO"
.IR tlspool-daemon "(8), " tlspool_starttls "(3), " pavlov "(3)."
.PP
Online resources may be found on the project home page,
.IR http://tlspool.arpa2.net .


tool/tlstunnel.c
/* tlspool/tlstunnel.c -- Simple utility, forward TLS as TCP and vice versa */

#include <stdlib.h>
#include <stdio.h>
#include <stdio.h>
#include <string.h>
#include <pthread.h>
#include <fcntl.h>

#include <errno.h>
#include <poll.h>
#include <unistd.h>
#include <getopt.h>
#include <netdb.h>

#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/wait.h>
#include <sys/select.h>
#include <sys/resource.h>

#include <tlspool/starttls.h>
#include <arpa2/pavlov.h>


static starttls_t tlsdata = {
	.flags = 0,
	.local = 0,
};

static char *pavlovcommand;
static char *tunnelcommand;
static int global_argc;
static char **global_argv;
static struct addrinfo *remoteaddrinfo;
static int role = -1;

static struct option choice_clisrv [] = {
	{ "client", 0, NULL, 'c' },
	{ "server", 0, NULL, 's' },
	{ NULL, 0, NULL, 0 }
};

static struct option command_options [] = {
	{ "tcp",		0, NULL, 't' },
	{ "tcp-tls",		0, NULL, 't' },
	{ "udp",		0, NULL, 'u' },
	{ "udp-dtls",		0, NULL, 'u' },
	{ "sctp-stream-number",	1, NULL, 'x' },
	{ "sctp-tunneled",	0, NULL, 'o' },
	{ "sctp-tunnelled",	0, NULL, 'o' },
	{ "sctp-not-tunneled",	0, NULL, 'O' },
	{ "sctp-not-tunnelled",	0, NULL, 'O' },
	{ "sctp-dtls",		0, NULL, 'd' },
	{ "sctp-tls",		0, NULL, 'D' },
	{ "fork",		0, NULL, 'f' },
	{ "service",		1, NULL, 's' },
	{ "local-addr",		1, NULL, 'l' },
	{ "remote-addr",	1, NULL, 'r' },
	{ "local-id",		1, NULL, 'L' },
	{ "remote-id",		1, NULL, 'R' },
	{ "tlspool-socket-path",1, NULL, 'S' },
	{ "pavlov-command",	1, NULL, 'p' },
	{ NULL, 0, NULL, 0 }
};



#ifdef USING_STR2PORT_SOMEWHERE
int str2port (char *portstr) {
	char *portrest;
	long int retval = strtol (portstr, &portrest, 0);
	if (*portrest) {
		fprintf (stderr, "Not a port number: %s\n", portstr);
		exit (1);
	}
	if ((retval <= 0) || (retval > 65535)) {
		fprintf (stderr, "Port numbers range from 1 to 65535: %s\n", portstr);
		exit (1);
	}
	return (int) retval;
}
#endif


struct unixaddrinfo {
	struct addrinfo ai;
	struct sockaddr_un sa;
};

void parse_addrinfo (char *instr, char local_remote,
			struct addrinfo *hint, struct addrinfo **res) {
	while (*res != NULL) {
		res = &(*res)->ai_next;
	}
	if (!instr) {
		fprintf (stderr, "Specify -%c as either a socket pathname or a [host]:port\n", local_remote);
		exit (1);
	}
	if (instr [0] == '/') {
		struct unixaddrinfo *uai;
		/* Interpret instr as a UNIX domain socket's path name */
		uai = malloc (sizeof (struct unixaddrinfo));
		if (uai == NULL) {
			fprintf (stderr, "Out of memory allocating UNIX addrinfo\n");
			exit (1);
		}
		uai->ai.ai_flags = 0;
		uai->ai.ai_family = AF_UNIX;
		uai->ai.ai_socktype = hint->ai_socktype;
		uai->ai.ai_protocol = 0;
		uai->ai.ai_addrlen = sizeof (struct sockaddr_un);
		uai->ai.ai_addr = (struct sockaddr *) &uai->sa;
		uai->ai.ai_canonname = instr;
		uai->ai.ai_next = NULL;
		uai->sa.sun_family = AF_UNIX;
		if (strlen (instr) +1 > sizeof (uai->sa.sun_path)) {
			fprintf (stderr, "UNIX domain socket path %s too long\n", instr);
			exit (1);
		}
		strcpy (uai->sa.sun_path, instr);
		*res = &uai->ai;
	} else {
		/* Interpret instr as a host:port combination */
		char *port;
		int braket;
		port = strrchr (instr, ':');
		if (port == NULL) {
			fprintf (stderr, "Format -%c as either /unix/domain/path/name or as host:port\n", local_remote);
			exit (1);
		}
		if ((instr [0] == '[') && (port [-1] == ']')) {
			braket = 1;
			hint->ai_family = AF_INET6;
		} else {
			braket = 0;
			hint->ai_family = AF_UNSPEC;
		}
		port [-braket] = '\0';
		if (getaddrinfo (instr + braket, port+1, hint, res) != 0) {
			fprintf (stderr, "Syntax error in -%c paramater; use /uinx/domain/path/name or host:port\n", local_remote);
			exit (1);
		}
		port [-braket] = braket? ']': ':';
	}
}


int run_pavlov (int plainfd) {
	int pavlovexit = 0;
	//
	// Without explicit pavlovcommand, invoke pavlov scripting for expect/response
	if (pavlovcommand == NULL) {
		return pavlov (plainfd, plainfd, tunnelcommand, global_argc, global_argv);
	}
	//
	// Fork off a STARTTLS pavlov session
	switch (fork ()) {
	case -1:
		perror ("Failed to fork pavlov command");
		return 1;
	case 0:
		close (0);
		close (1);
		if ((dup2 (plainfd, 0) == -1) || (dup2 (plainfd, 1) == -1)) {
			fprintf (stderr, "Failed to connect plaintext to the %s interaction\n", tunnelcommand);
			exit (2);
		}
		fprintf (stderr, "Starting script + args: %s, %s, %s...\n", tunnelcommand, global_argv [1], global_argv [2]);
		execve (pavlovcommand, global_argv, NULL /*TODO:envp:params*/);
		perror ("Failed to start alternative to the pavlov command");
		exit (2);
	default:
		wait (&pavlovexit);
		if (!WIFEXITED (pavlovexit)) {
			return 2;
		}
		return WEXITSTATUS (pavlovexit);
	}
}



/* Format a UNIX domain socket path that may contain formatting characters:
 *
 * %L and %*L indicate the localid  negotiated in the TLS handshake;
 * %R and %*R indicate the remoteid negotiated in the TLS handshake;
 *
 * Note that a server has authenticated names, but a client does not, due
 * to the early stage in which it forms these addresses (before the TLS
 * handshake).
 *
 * The * is ignored for now, but may be used to prefix domain patterns.
 *
 * There currently is no support for iterators with DoNAI selectors, but
 * this is a logical extension (this function would recurse).
 */
int fmtcpy (char *dst, const char *src, size_t dstsz, starttls_t *tlsdata) {
	while (src && *src) {
		size_t len;
		char *perc = strchr (src, '%');
		if (perc) {
			len = ((intptr_t) perc) - ((intptr_t) src);
		} else {
			len = strlen (src);
		}
		if (len + 1 > dstsz) {
			return -1;
		}
		memcpy (dst, src, len);
		src   += len;
		dst   += len;
		dstsz -= len;
		if (perc) {
			int star = *(++src) == '*';
			char *xtra = "";
			if (star) {
				src++;
			}
			switch (*src++) {
			case 'L':
				xtra = tlsdata->localid;
				break;
			case 'R':
				xtra = tlsdata->remoteid;
				break;
			default:
				return -1;
			}
			len = strnlen (xtra, 127);
			if (len + 1 > dstsz) {
				return -1;
			}
			memcpy (dst, xtra, len);
			dst   += len;
			dstsz -= len;
		}
	}
	*dst++ = '\0';
	return 0;
}


//TODO// Spark thread with:  fd, localaddrinfo, role, remoteaddrinfo, pavlovcommand, tlsdata, tunnelcommand, argc, argv; only dynamic is fd, localaddrinfo; thread negotiates session and then goes down; curtlsdata is a thread local variable

/* Data structure passed to connection handler threads */
struct fdinfo {
	int fd;
	struct addrinfo *localaddrinfo;
};

/* We will smuggle a file descriptor into the localaddrinfo field ai_family;
 * the value in ai_family is no longer needed, and it would be replicated in
 * ai_addr->sa_family as well as available through the getsockaddr() result.
 */
#define smuggle_fd ai_family


#ifdef NEED_THIS_WHEN_THE_CONNECTION_CLOSES_FINALLY
void cleanup_connection (void *vfdi) {
	struct fdinfo *fdi = vfdi;
	struct addrinfo *lai = fdi->localaddrinfo;
	if (lai->smuggle_fd < 0) {
		/* Correct UDP reversal so it will be selected again */
		lai->smuggle_fd = -lai->smuggle_fd;
	}
	if (fdi->fd != lai->smuggle_fd) {
		/* For TCP and SCTP, close the connection */
		close (fdi->fd);
	}
}
#endif


/* The connect_remote() call finds a remoteaddrinfo of the same ai_socktype
 * as the vlai argument; vlai is localaddrinfo but is passed as (void *) to
 * support its use in starttls_xxx as a connect_plaintext() routine.  This
 * is also the reason for the starttls_t data (that is being ignored).
 *
 * The return value is -1 on error, and errno will then be set.  Successful
 * return is a file descriptor >= 0.
 */
int connect_remote (starttls_t *curtlsdata, void *vlai) {
	struct addrinfo *lai = vlai;
	struct addrinfo *rai = remoteaddrinfo;
	while (rai) {
		if (rai->ai_socktype == lai->ai_socktype) {
			struct sockaddr_un sun;
			struct sockaddr *sai = rai->ai_addr;
			int sailen = rai->ai_addrlen;
			int sox = socket (sai->sa_family, rai->ai_socktype, rai->ai_protocol);
			if (sox == -1) {
				return -1;
			}
			if (sai->sa_family == AF_UNIX) {
				memset (&sun, 0, sizeof (sun));
				sun.sun_family = AF_UNIX;
				const char *src_path = ((struct sockaddr_un *) sai)->sun_path;
				if (fmtcpy (sun.sun_path, src_path, sizeof (sun.sun_path), curtlsdata) != 0) {
					fprintf (stderr, "Formatted socket path too long or badly formatted: %s\n", src_path);
					close (sox);
					continue;
				}
				fprintf (stderr, "DEBUG: Formatting returned %s\n", sun.sun_path);
				sai = (struct sockaddr *) &sun;
			}
			if (connect (sox, sai, sailen) == 0) {
				return sox;
			}
			close (sox);
		}
		rai = rai->ai_next;
	}
	return -1;
}


/* The connection_thread() uses the facilities of starttls_xxx() to
 * request the plaintext fd.  It also runs pavlov() over the cryptfd before
 * invoking starttls_xxx(), as a preamble to the encrypted link.
 *
 * When acting as a client:
 *  - local initiator cnx is plainfd
 *  - fwd/cryptfd is constructed early, using connect_remote()
 *  - fwd/cryptfd is then subjected to pavlov()
 *  - fwd/cryptfd is then passed into starttls_client()
 *  - cnx/plainfd is setup in int *privdata, connect_plaintext is NULL/default
 *
 * When acting as a server:
 *  - local intitiator cnx is cryptfd
 *  - cnx/cryptfd first subjected to pavlov()
 *  - cnx/cryptfd is that passed into starttls_server()
 *  - fwd/plainfd is constructed on demand, in connect_plaintext()
 *  - fwd/plainfd construction is connect_remote() with privdata==localaddrinfo
 */

void *connection_thread (void *vfdi) {
	struct fdinfo *fdi = vfdi;
	int cnx = fdi->fd;
	int plainfd = -1;
	int cryptfd = -1;
	//TODO// pthread_cleanup_push (cleanup_connection, vfdi);
	printf ("Thread started to handle fd=%d\n", cnx);
	int fwd;
	int setup;
	starttls_t curtlsdata = tlsdata;	/* local working copy */

	//
	// Setup cnx/fwd and/or cryptfd/plainfd, inasfar as possible now
	if (role == 's') {
		cryptfd = cnx;
	} else {
		plainfd = cnx;
		fwd = cryptfd = connect_remote (&curtlsdata, fdi->localaddrinfo);
	}
	//
	// Run pavlov on the cryptfd, as a preamble to starttls
	setup = run_pavlov (cryptfd);
	if (setup != 0) {
		fprintf (stderr, "Failed pavlovian precursor to TLS, error code %d\n", setup);
		return NULL;
	}
	//
	// Invoke starttls
	if (role == 's') {
		// Server: on-demand connect_remote based on localaddrinfo
		setup = tlspool_starttls (cryptfd, &curtlsdata, fdi->localaddrinfo, connect_remote);
	} else {
		// Client: plainfd already available, default returns it
		setup = tlspool_starttls (cryptfd, &curtlsdata, &plainfd, NULL);
	}
	if (setup == -1) {
		perror ("Failed to start TLS");
		if (plainfd >= 0) {
			close (plainfd);
			plainfd = -1;
		}
	}
	//
	// Cleanup, as the TLS Pool now connects the end points
	if (plainfd >= 0) {
		close (plainfd);
		plainfd = -1;
	}
	//TODO// pthread_cleanup_pop (1);
	return NULL;
}


int main (int argc, char *argv []) {
	int parsing = 1;
	int carrier = -1;
	int sctpdtls = 1;
	int sctpudp = 0;
	long parsed_number;
	int stream;
	int tlsfork = 0;
	char *localid = NULL;
	char *remotid = NULL;
	char *cmdsoxpath = NULL;
	struct addrinfo hint;
	struct addrinfo *localaddrinfo = NULL;
	struct addrinfo *addrwalk = NULL;
	fd_set bindings, rselect;
	int maxbound;
	int sox = -1;
	int argc_skip;
	int cnxlim;
	struct rlimit rlimit_nofile;
	struct fdinfo *fdmem;

	//
	// Initialise variables
	tunnelcommand = argv [0];
	//
	// Fetch the limitation on the number of conncetions
	if (getrlimit (RLIMIT_NOFILE, &rlimit_nofile) == -1) {
		perror ("Failed to extract file number limit");
		exit (1);
	}
	//
	// Parse the command line arguments
	memset (&hint, 0, sizeof (hint));
	hint.ai_family = AF_UNSPEC;
	hint.ai_socktype = SOCK_STREAM;
	//
	// First option is either -c for client or -s for server
	//TODO// Future options may include peering
	switch (getopt_long (argc, argv, "cs", choice_clisrv, NULL)) {
	case 'c':
		/* -c for client */
		role = 'c';
		tlsdata.flags = PIOF_STARTTLS_LOCALROLE_CLIENT
				| PIOF_STARTTLS_REMOTEROLE_SERVER;
		break;
	case 's':
		/* -s for server */
		role = 's';
		tlsdata.flags = PIOF_STARTTLS_LOCALROLE_SERVER
				| PIOF_STARTTLS_REMOTEROLE_CLIENT;
		break;
	case -1:
	default:
		break;
	}
	if (role == -1) {
		fprintf (stderr, "The first argument should be -c or -s to setup a TLS client or TLS server\n");
		exit (1);
	}
	//
	// Further options are settings
	while (parsing) {
		//TODO// getlongopt
		//TODO// -d for DTLS / -D for TLS; -w for SCTP-over-UDP; -W not
		int opt = getopt_long (argc, argv,
			"udDtx:oO:fs:l:r:L:R:S:C:",
			command_options, NULL);
		switch (opt) {
		case 'u':
			/* -u for DTLS/UDP */
			hint.ai_socktype = SOCK_DGRAM;
			fprintf (stderr, "UDP mode wrapping with DTLS is not implemented -- see man page\n");
			exit (1);
		case 'd':
			/* Run SCTP with DTLS (default) */
			sctpdtls = 1;
			break;
		case 'D':
			/* Run SCTP with TLS */
			sctpdtls = 0;
			break;
		case 'o':
			/* Run SCTP over an UDP tunnel */
			sctpudp = 1;
			break;
		case 'O':
			/* Run SCTP directly (default) */
			sctpudp = 0;
			break;
		case 't':
			/* Run over TCP (default) */
			hint.ai_socktype = SOCK_STREAM;
			break;
		case 'x':
			/* DTLS/SCTP, DTLS/SCTP/UDP, TLS/SCTP or TLS/SCTP/UDP */
			//TODO// Maybe better to use a comma-separated list
			parsed_number = strtol (optarg, &optarg, 0);
			if (*optarg) {
				fprintf (stderr, "Syntax error in stream number to -%d\n", opt);
				exit (1);
			}
			if ((parsed_number < 0) || (parsed_number > 65535)) {
				fprintf (stderr, "Stream numbers range from 0 to 65535\n");
				exit (1);
			}
			hint.ai_socktype = SOCK_SEQPACKET;
			stream = parsed_number;
			break;
		case 'f':
			/* -f to fork TLS sessions */
			if (tlsfork) {
				fprintf (stderr, "You should only specify TLS session forking once\n");
				exit (1);
			}
			tlsfork = 1;
			break;
		case 's':
			/* -s servicename */
			if (*tlsdata.service) {
				fprintf (stderr, "You should only specify one service name\n");
				exit (1);
			}
			if (strlen (optarg) > TLSPOOL_SERVICELEN - 1) {
				fprintf (stderr, "Service names should not exceed a length of %d characters (and be IANA-defined)\n", TLSPOOL_SERVICELEN - 1);
				exit (1);
			}
			strcpy ((char *)(&tlsdata.service[0]), optarg);
			break;
		case 'l':
			/* -l xxx for local  address xxx */
			parse_addrinfo (optarg, 'l', &hint, &localaddrinfo);
			break;
		case 'r':
			/* -l xxx for remote address xxx */
			parse_addrinfo (optarg, 'r', &hint, &remoteaddrinfo);
			break;
		case 'L':
			/* -L xxx for localid xxx */
			if (*tlsdata.localid) {
				fprintf (stderr, "For now, it is not permitted to specify multiple local identities\n");
				exit (1);
			}
			if (1 + strlen (optarg) > sizeof (tlsdata.localid)) {
				fprintf (stderr, "Local identity is too long\n");
				exit (1);
			}
			strcpy (tlsdata.localid, optarg);
			break;
		case 'R':
			/* -R xxx for remoteid xxx */
			if (*tlsdata.remoteid) {
				fprintf (stderr, "Cannot constrain to multiple remote identities at the same time\n");
				exit (1);
			}
			if (1 + strlen (optarg) > sizeof (tlsdata.remoteid)) {
				fprintf (stderr, "Remote identity is too long\n");
				exit (1);
			}
			strcpy (tlsdata.remoteid, optarg);
			break;
		case 'S':
			/* -S /sox/path for TLS Pool socket path /sox/path */
			if (cmdsoxpath) {
				fprintf (stderr, "You can specify only one TLS Pool command socket path\n");
				exit (1);
			}
			cmdsoxpath = strdup (optarg);
			break;
		case 'p':
			/* -p /usr/bin/pavlov for external pavlov replacement */
			if (pavlovcommand) {
				fprintf (stderr, "You can specify only one replacement for the pavlov command\n");
				exit (1);
			}
			pavlovcommand = strdup (optarg);
			break;
		case -1:
			parsing = 0;
		}
	}
	//
	// Sanity checks
	if (role == -1) {
		fprintf (stderr, "Specify either -c or -s (client or server mode)\n");
		exit (1);
	}
	if (!*tlsdata.localid) {
		fprintf (stderr, "You need to specify -L (localid)\n");
		exit (1);
	}
	if (!localaddrinfo) {
		fprintf (stderr, "You need to specify -l (local socket address)\n");
		exit (1);
	}
	if (!remoteaddrinfo) {
		fprintf (stderr, "You need to specify -r (remote socket address)\n");
		exit (1);
	}
	if (cmdsoxpath) {
		if (tlspool_open_poolhandle (cmdsoxpath) == INVALID_POOL_HANDLE) {
			perror ("Failed to open TLS Pool command socket");
			exit (1);
		}
	}
	tlsdata.flags |= (sctpdtls? PIOF_STARTTLS_DTLS: 0);	//TODO// Later
	if (role == 'c') {
		//DO-SEND-SNI// tlsdata.flags |= PIOF_STARTTLS_WITHOUT_SNI;
	}
	if (tlsfork) {
		tlsdata.flags |= PIOF_STARTTLS_FORK;
	}
	//
	// Allocate an information structure per file descriptor
	fdmem = malloc (sizeof (struct fdinfo [rlimit_nofile.rlim_cur]));
	if (fdmem == NULL) {
		fprintf (stderr, "Failed to allocate connection data structures\n");
		exit (1);
	}
	//
	// Collect the pavlov arguments
	global_argv = argv + (optind - 1);   /* First is arg, not progname! */
	global_argc = argc - (optind - 1);
#ifdef DEBUG
	printf ("argv_pavlov = %s, %s, %s... (skipped %d)\n", argv [argc_skip], argv [argc_skip+1], argv [argc_skip+1], argc_skip);
#endif
	//
	// Listen to the incoming address
	//TODO// Setup localaddrinfo/remoteaddrinfo ai_socktype as a "hint"
	//TODO// Sockets are opened in the binding loop, using localaddrinfo
	FD_ZERO (&bindings);
	maxbound = 0;
	addrwalk = localaddrinfo;
	while (addrwalk) {
		int true = 1;
		long fcntl_flags;
		int sox = socket (addrwalk->ai_family, addrwalk->ai_socktype, addrwalk->ai_protocol);
		if (sox == -1) {
			fprintf (stderr, "Failed to create socket for %s: %s\n",
					addrwalk->ai_canonname,
					strerror (errno));
			exit (1);
		}
		// Share the socket; used because the TLS Pool holds older
		// connections alive while restarting a TLS Tunnel.
		if (setsockopt (sox, SOL_SOCKET, SO_REUSEADDR, &true, sizeof (true)) != 0) {
			fprintf (stderr, "Failed to setup socket for reuse of local address (non-fatal)\n");
		}
		// Set the socket to non-blocking mode; this avoids a lockup
		// on accept() when select() reports a connection attempt that
		// is retracted before accept() is tried.
		fcntl_flags = fcntl (sox, F_GETFL, 0);
		if (fcntl_flags >= 0) {
			if (fcntl (sox, F_SETFL, fcntl_flags | O_NONBLOCK) != 0) {
				fcntl_flags = -1;
			}
		}
		if (fcntl_flags < 0) {
			fprintf (stderr, "Failed to setup for non-blocking accept()\n");
		}
		if (bind (sox, addrwalk->ai_addr, addrwalk->ai_addrlen) == -1) {
			fprintf (stderr, "Failed to bind to %s: %s\n",
					addrwalk->ai_canonname,
					strerror (errno));
			exit (1);
		}
		if (addrwalk->ai_socktype != SOCK_DGRAM) {
			if (listen (sox, 5) == -1) {
				fprintf (stderr, "Failed to listen to %s: %s\n",
						addrwalk->ai_canonname,
						strerror (errno));
				exit (1);
			}
		}
		FD_SET (sox, &bindings);
		if (sox > maxbound) {
			maxbound = sox;
		}
		addrwalk->smuggle_fd = sox; // Note: destroys ai_family info!!
		addrwalk = addrwalk->ai_next;
	}
	//
	// Fork off a daemon process
	switch (fork ()) {
	case -1:
		perror ("Failed to fork background daemon");
		exit (1);
	case 0:
		if (setsid () == (pid_t) -1) {
			perror ("Failed to create background session");
			exit (1);
		}
		break;
	default:
		printf ("DEBUG: Forked daemon process to the background\n");
		exit (0);
	}
	//
	// Handle incoming connections
	//TODO// Support simultaneous connections for real-life tunnel use
	while (1) {
		int cnx;
		rselect = bindings;
		if (select (maxbound+1, &rselect, NULL, NULL, NULL) == -1) {
			perror ("Failed to select()");
			//TODO// Is there a reason for recovery?
			exit (1);
		}
		addrwalk = localaddrinfo;
		while (addrwalk) {
			if ((addrwalk->smuggle_fd >= 0) && FD_ISSET (addrwalk->smuggle_fd, &rselect)) {
				pthread_t thr;
				if (addrwalk->ai_socktype == SOCK_DGRAM) {
					// UDP handling must be delegated
					// completely and no longer polled here
					cnx = addrwalk->smuggle_fd;
					addrwalk->smuggle_fd = -addrwalk->smuggle_fd;
				} else {
					cnx = accept (addrwalk->smuggle_fd, NULL, 0);
					if (cnx == -1) {
						// Spurious RST detected, ignore
						continue;
					}
				}
				//
				// Spark a thread to handle to connection
				fdmem [cnx].fd = cnx;
				fdmem [cnx].localaddrinfo = addrwalk;
				pthread_create (&thr, NULL, (void *) connection_thread, (void *) &fdmem [cnx]);
				pthread_detach (thr);
			}
			addrwalk = addrwalk->ai_next;
		}
	}
	exit (1);
}

